### PR TITLE
Multiple file systems

### PR DESCRIPTION
**Pull request by felixvisee**
Jira issue: [WP-44](http://jira.webinos.org/browse/WP-44), [WP-601](http://jira.webinos.org/browse/WP-601), [WP-694](http://jira.webinos.org/browse/WP-694)
## Overview
- Zap and Shake/Slide and Share demo hack solved ([WP-44](http://jira.webinos.org/browse/WP-44)).
  - `resolveLocalFileSystemURL` and `toURL` are no longer supported
  - `getLink` asynchronously generates a temporary URL (`FileEntry` only).
- Support for multiple file systems ([WP-601](http://jira.webinos.org/browse/WP-601)).
  - Including support for local file systems and dropbox file systems.
  - Configuration in `webinos_config.json` (subject to changes).
- Remove unnecessary dependencies ([WP-694](http://jira.webinos.org/browse/WP-694)).
  - Removed `inherits`, added `express@3.1.x`, `node-uuid@1.4.x`, and `dbox@0.5.x`.
  - Keeping `mkdirp@0.3.x` for now.
## Shares

We are now introducing the concept of _shares_, i.e., a shared folder on one of the supported file systems (currently local and dropbox). At the moment, shares are configured in the `webinos_config.json`. This will change however, when opinions about configuration of webinos eventually converge.
### Local shares

Sharing a folder from the local file system requires creating a named share including the full path to the folder. The default `webinos_config.json` includes an example (excerpt):

``` json
    { "name" : "file"
    , "params" :
      { "local" :
        { "shares" :
          [ { "name" : "desktop"
            , "path" : "/path/to/desktop"
            }
          ]
        }
      }
    }
```

Generating temporary URLs and serving underlying content from the local file system also requires setting up a server (default: `localhost:9999`):

``` json
    { "name" : "file"
    , "params" :
      { "local" :
        { "server" :
          { "port" : 9999
          , "hostname" : "localhost"
          }
        }
      }
    }
```
### Dropbox shares

In addition to creating a named share including the full path to the folder rooted at the webinos app folder, sharing a folder from a dropbox file system requires an access token, which can be obtained as follows (copied from [sintaxi/node-dbox](https://github.com/sintaxi/node-dbox)).
#### Step 1

Create an `app` using application credentials provided by dropbox.

``` javascript
var dbox = require("dbox")
var app = dbox.app({ "app_key": "38j04m9q86j2zva", "app_secret": "pzl7f411nu6dy4p" })
```
#### Step 2

Authorization is a three step process.

a) Get a request token.

``` javascript
var request_token
app.requesttoken(function (status, value) {
  request_token = value
})
```

b) Visit the URL to grant authorization.

`https://www.dropbox.com/1/oauth/authorize?oauth_token=#{ request_token.oauth_token }`

c) Generate the access token with the request token.

``` javascript
app.accesstoken(request_token, function (status, access_token) {
  console.log(access_token)
})
```
#### Step 3

Configure webinos to use the obtained access token and set up the desired shares (excerpt):

``` json
    { "name" : "file"
    , "params" :
      { "dropbox" :
        { "access_token" :
          { "oauth_token_secret" : "a12bc3d4e5fghi6"
          , "oauth_token" : "ab1c2defghijklm"
          , "uid" : "1234567"
          }
        , "shares" :
          [ { "name" : "full"
            , "path" : "/"
            }
          ]
        }
      }
    }
```

Generating temporary URLs and serving underlying content from a dropbox file system is delegated to dropbox.
## Security & Policies

Instead of enforcing any origin-specific security checks as originally specified (see [File API: Directories and System](http://www.w3.org/TR/file-system-api/)), access control is delegated to the Policies API. Unfortunately, persistently attaching policies to dynamic services is an open issue. Further investigation is required.

### CODE REVIEW

**Comment by felixvisee** [comment]
Hm, webinos app at dropbox is currently limited to 5 users. We have to apply for production status to raise (remove?!) this limit. We may also want to create a webinos account and attach the app to it instead of using my personal account?!

File: package.json [hunk 1]
```diff
@@ -26,9 +26,8 @@
   , "now" : "0.8.1"
   , "node-uuid" : "1.3.x"
   , "find-dependencies" : "~0.0.2"
-  , "inherits" : "1.0.0"
   , "mkdirp" : "0.3.x"
-  , "connect" : "2.6.x"
+  , "dbox": "https://github.com/felixjendrusch/node-dbox/tarball/put-zero"
   , "grunt-contrib-clean" : "~0.3.x"
```
Comment (cbotsikas) [inline package.json]: You will need to rebase your package.json, grunt-contrib-clean has been moved on a resent PR merge

**Comment by andreasbotsikas** [comment]
Great job @felixjendrusch. For the windows installer, the end user would expect to install webinos and run it as is without having to configure the location of the initial file api. Do you think it is possible to have a configuration variable that reads something like $webinosLocalFilePath$ which would internally be translated to the webinosPath ~/home/.webinos/pzp folder, %appdata%\roaming\webinos\pzp or /mnt/sdcard/.webinos? @cbotsikas has mentioned that it is possible to use the https://github.com/webinos/Webinos-Platform/blob/master/webinos/core/util/lib/webinosPath.js file inside your module.

**Comment by mlasak** [comment]
I agree, initially there should be as well no dropbox-share defined to avoid confusion having a useless service, that only responds with error callbacks.

How about not sharing anything by default and letting the user to decide even on the "initial" share. This would drive the service configuration discussion, as there would be no file service until explicitly configured so by the user.

**Comment by andreasbotsikas** [comment]
This sounds like a nice idea (although it will be tough for the devs). But then, the service should not be activated unless the end use has selected a file share, right? This will also require a "suggest the location" button in the ui cause we shouldn't expect the end users to set the /mnt/sdcard path for android to work... This is why I initially said to set this by default instead of forcing the user to set it.

File: package.json [hunk 2]
```diff
@@ -27,12 +27,11 @@
   , "now" : "0.8.1"
   , "node-uuid" : "1.3.x"
   , "find-dependencies" : "~0.0.2"
-  , "inherits" : "1.0.0"
   , "mkdirp" : "0.3.x"
-  , "connect" : "2.6.x"
+  , "dbox": "https://github.com/felixjendrusch/node-dbox/tarball/put-zero"
```
Comment (jplyle) [inline package.json:31]: Could this be an optional dependency?

**Comment by jplyle** [comment]
I've only just checked this all out.  The idea gets a firm +1 from me.

However, I think having no default share is a mistake.  it means our File API test on the test bed doesn't work, and it'll break things for anyone developing for webinos.  It's particularly important because there isn't a usable tool to create new shares yet.

How does this work for default application storage areas?  I would assume that the app storage share (~/.webinos/userData/appname/ or similar) would be available by default, but that the policy manager would prevent it from being accessible to applications unless they have the required permission.  Additional shares should be configurable and created based on user interaction, I agree.

I can confirm that this pull request doesn't appear to break anything - I've tested enrolment, inter-user certificate exchange and service sharing all on Linux.

When I run the testbed, despite modifying webinos_config.json to include the local share as per the example, I can't get the File API demo to work.  Has this been updated?

A test or two wouldn't go amiss, either... the Jasmine API tests are failing because the service isn't discoverable.

**Comment by jplyle** [comment]
Salvo has pointed out that my issues with the file API share are probably due to me not deleting my .webinos directory between tests.

**Comment by habibvirji** [comment]
@felixjendrusch please fix the above raised issue by John.

**Comment by mlasak** [comment]
The tasks for a  "tool to create new shares" and the service configuration is not / should not be part of the specific file api implementation. Service configuration is a separate more generic question to be discussed. Nevertheless the  tests and testbed page need to be updated, e.g. have left view for one service and right view for another. Thx @felixjendrusch for the great work.

**Pull request merged**